### ISSUE TITLE
Race condition in CloudSdk.getVersion

### ISSUE THREAD

**Issue by elharo** [GoogleCloudPlatform/appengine-plugins-core#269]
Tentative hypothesis: CloudSdk.getVersion sometimes returns before the stdOut line listener thread has started. This causes an IllegalArgumentException.

1. CloudSdk.getVersion calls runSynchronousGcloudCommand("gcloud info --format="value(basic.version)")
2. ProcessRunner runner = new DefaultProcessRunner(various args including the listeners)
3. runner.run(command.toArray(new String[command.size()]));
4. processBuilder.command(command);
    Process process = processBuilder.start();
5. handleStdOut(process);
6. stdOutThread.start();
7. syncRun(process)

  At this point the gcloud info command is off and running, and can return at any time. 
  However I see no guarantee that the thread started in step 6 has actually kicked off and is listening before the process from step 7 finishes. I think sometimes the process finishes before the stdOut thread starts, and thus we get no output on stdOut, which returns an empty string as the version and thus causes an IllegalArgumentException.

**Comment by elharo** [comment]
Here's my proposal to fix this:

Ditch the listener streams.


  public DefaultProcessRunner(boolean async,
                              List<ProcessExitListener> exitListeners,
                              List<ProcessStartListener> startListeners,
                              List<ProcessOutputLineListener> stdOutLineListeners,
                              List<ProcessOutputLineListener> stdErrLineListeners) {
    this(async, exitListeners, startListeners, false /* inheritProcessOutput */);
    this.stdOutLineListeners.addAll(stdOutLineListeners);
    this.stdErrLineListeners.addAll(stdErrLineListeners);
  }

getVersion is the only method that uses stdOutLineListeners and it only supplies one. Nothing uses stdErrLineListeners as far as I can see.

Instead, redirect stdOut into a file and then  read from that when we're done. Maybe even bypass ProcessRunner completely.

**Comment by elharo** [comment]
Also possible: the main thread gets all the way to the end of runSynchronousGCloudCommand where it calls

    return stdOutListener.toString();

before stdOutThread has called   stdOutLineListener.onOutputLine(line); once.

In fact, I think that's the more likely case.

**Comment by chanseokoh** [comment]
The hypothesis seems plausible.

I think it's not just `getVersion()` that supplies the listeners. FWIW, our tools as a client of the core library supply `stdOutLineListeners` and `stdErrLineListeners`. For example, in Eclipse:

https://github.com/GoogleCloudPlatform/google-cloud-eclipse/blob/master/plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerBehaviour.java#L295

Maybe it will be easier to sync with (i.e., wait for) the stdout handler thread and the stderr handler thread before calling `syncRun()` or `asyncRun()`.

**Comment by elharo** [comment]
And here's a third possibility. StringBuilders are not thread safe. There's no guarantee that the StringBuilder object in the stdOut thread has been updated by the time we call stdOutListener.toString()

**Comment by chanseokoh** [comment]
I agree on all the comments of @elharo. We should also sync with the termination of the stdout/stderr handler threads and also the shared object should be accessed in a thread-safe way.

**Comment by elharo** [comment]
This is going to be tricky because the stdout/stderr handler threads don't terminate. They're daemon threads and Scanner#hasNextLine blocks:

      public void run() {
        while (stdOut.hasNextLine() && !Thread.interrupted()) {
          String line = stdOut.nextLine();
          for (ProcessOutputLineListener stdOutLineListener : stdOutLineListeners) {
            stdOutLineListener.onOutputLine(line);
          }
        }
        stdOut.close();
      }

**Comment by elharo** [comment]
Even keeping the API, the safest approach might still be to redirect the output into a file and then stream it into the various listeners a line at a time after the process completes, all in a single, sequential thread. 

Are any gcloud commands expected to generated 100,000+ lines of output?

**Comment by chanseokoh** [comment]
Both `Scanner.hasNextLine()` and `Scanner.nextLine()` throws `IllegalStateException` if the scanner is closed, according to the [javadoc](https://docs.oracle.com/javase/7/docs/api/java/util/Scanner.html#hasNextLine()). My understanding is that, if a process is terminated (for whatever reasons), every open file descriptor of the process is closed and cleaned up by OS. So, I *think* the handler threads will eventually terminate too when the process terminates.

However, due to race condition, we still need to sync with the termination of the handler threads even though the main thread currently waits for the termination of the process (via `process.waitFor()`).

**Comment by chanseokoh** [comment]
> .. redirect the output into a file ...

Thinking for a while, I don't find a real difference from outputting to a "in-memory file" (i.e., buffer).

> and then stream it into the various listeners a line at a time after the process completes

And we can still do the same thing with the in-memory buffer. For the `getVersion()` scenario, the output is a single line, so we just need to ensure via proper synchronization that `return stdOutListener.toString()` happens after the process (as well as the handler) completes, as you suggested. If we ensure this sync, then I think we don't even need to make the buffer thread-safe.

However, for other general gcloud processes, I think we still want to stream the output rather than we wait for process termination. For example, deploy could take several minutes to terminate.

**Comment by chanseokoh** [comment]
>> .. redirect the output into a file ...

> Thinking for a while, I don't find a real difference from outputting to a "in-memory file" (i.e., buffer).

Sorry, thinking again, there does exist a difference.

**Comment by chanseokoh** [comment]
Yeah, I think checking the output of a process through this stdout/stderr handler mechanism is inherently unsafe. As an extreme example,

1. The stdout handler and a process start together.
1. The process makes some output.
1. However, the handler has not yet opened the input stream of the process or read (some or all of) the output lines.
1. The process terminates.
1. The handler attempts to open the input stream or read a line. Since the process is already terminated, the handler will be terminated with an exception.
1. As a result, we couldn't read any output of the process.

Basically, there is no guarantee that the handler will read all output lines, even if we sync with the process and handler start/termination.

**Comment by briandealwis** [comment]
> Also possible: the main thread gets all the way to the end of runSynchronousGCloudCommand where it calls
> return stdOutListener.toString();

Yeah, that's it: it needs to join on the `stdOutThread` and `stdErrThread`.

> This is going to be tricky because the stdout/stderr handler threads don't terminate. They're daemon threads and Scanner#hasNextLine blocks

When a process dies, it causes EOFs on `stdout` and `stderr`, and `hasNextLine()` should return false (see `Scanner#readInput()`).

**Comment by chanseokoh** [comment]
About https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues/269#issuecomment-264609748:

Looks like the JVM implementation (at least my JVM) actually buffers the output of a sub-process, so for a small output, I think the problem I described in that comment doesn't happen in practice. `hasNextLine()` and `nextLine()` works fine even if called well after a process dies. The following program prints out all the output:

```java
import java.io.IOException;
import java.util.Scanner;

public class A {

  public static void main(String[] args) throws IOException, InterruptedException {
    ProcessBuilder pb = new ProcessBuilder();
    pb.command("/bin/echo", "1\n2\n3\n4\n5");
    Process p = pb.start();
    p.waitFor();
    Thread.sleep(3000);

    Scanner s = new Scanner(p.getInputStream());
    while (s.hasNextLine()) {
      System.out.println("Read: " + s.nextLine());
    }
  }
}
```

So, for `getVersion()` where gcloud outputs a small single line, I think we can continue to use this handler mechanism, *as long as we sync with the handler termination.* (No need to wait for the handler to start.)

**Comment by elharo** [comment]
If anyone thinks they have a full fix for this, please send a PR. The synchronization here is going to be tricky.

If it is true that we have enough buffer to scan after the process exits, I'm tempted to do that in a single thread much like @chanseokoh's A example above, even if it requires setting up new methods just for `getVersion()`. I'd feel a lot more confident in the correctness of such a solution.

**Comment by elharo** [comment]
FYI, sync runs are only used privately within this library. I don't think Eclipse or IntelliJ can call them, so it's not out of the question to have a very different, simpler implementation for that case. 

asyncRuns probably still have this bug, and require a more complicated fix that synchronizes between threads properly.

**Comment by patflynn** [comment]
+1 to a signature change with a single threaded blocking model to getting the version.

**Comment by chanseokoh** [comment]
I will leave this open in case we want to make getVersion() a straight blocking call as in the above discussion.

**Comment by patflynn** [comment]
I'd open a new bug for that, since I'm hoping this particular race
condition issue is now fixed.
